### ISSUE TITLE
Recommend latin-1 for IRC server encoding?

### ISSUE THREAD

**Issue by Akuli** [ircdocs/ircdocs.horse#10]
https://modern.ircdocs.horse/#character-encodings says:

> For what it’s worth, the servers I’ve worked on have done the latter of the two options, discarding any lines that cannot be decoded as UTF-8. It’s not a great solution, but it’s a difficult issue to deal with. If anyone has a better way to deal with this, please feel free to contact me so I can update this section.

I am helping a friend write an IRC server, and when we were faced with this problem, I came up with a simple solution: use latin-1 everywhere in the server. This works as intended: whatever is sent to the server will come back unchanged, and decoding will never error. I explained this in more detail at https://github.com/ThePhilgrim/MantaTail/pull/64.

(I sent an email about this to DanielOaks, but later figured that a github issue would be more appropriate)

**Comment by progval** [comment]
Hi,

You are effectively suggesting to manipulate messages as raw bytes, but use `latin-1` encoding and decoding to shoehorn them into Python's `str` (aka unicode string) instead of `bytes`.

This is fine, as long as you don't need to do any text manipulation, such as truncation or concatenation; then you have the same issue whether you use raw bytes or `latin-1`.

Truncation cannot safely be done without knowledge of the encoding, and concatenating two strings with different encodings will result in mangled text in clients reading the line.

**Comment by progval** [comment]
Hmm I just read the context of the paragraph you quote, and it is indeed written to address the wish for high-level strings in Python-like languages rather than bytes.

So I think it should indeed be rephrased, but to recommend treating text as raw bytes more strongly (which is essentially equivalent to what you do with `latin-1`, but skipping the two transcoding steps). And for servers that really don't want to handle raw bytes, then we have a new solution since this was written: https://ircv3.net/specs/extensions/utf8-only (plus, they can do the same hybrid-encoding thingy as the clients)

**Comment by Akuli** [comment]
In practice, high level languages (at least Python and javascript) make it easier to work with strings than with bytes. For example, f-strings aren't available for bytes in python, so instead of `f"PRIVMSG {channel} :{message}"` you have to write `b"PRIVMSG %s :%s" % (channel, message)`. Not too bad, but it's a small annoyance for people who are used to writing f-strings, and the latin-1 workaround makes them work just as well as bytes.

**Comment by progval** [comment]
Yes, that's true. The right fix would be for Python to support "f-bytes", but that's never going to happen :/

**Comment by ksu-0** [comment]
The first and foremost problem with your proposal is that you are sending crap to clients.

You may also face issues if you casefold non-ASCII characters.

**Comment by progval** [comment]
> you are sending crap to clients

So they aren't. They `.decode('latin-1')` when receiving, then `.encode('latin-1')` before sending, as these two functions are reciprocal no matter the content; they even have a test for it here: https://github.com/ThePhilgrim/MantaTail/pull/64/files#diff-a845cc20d1de14382df57d38f99b0ac58af54751623253e2ed7cac20baf54b2eR544-R546

> You may also face issues if you casefold non-ASCII characters, or do string matching based on character classes.

that is true, however

**Comment by ksu-0** [comment]
They send data that the client does not know how to display. It can try to guess, but it is not possible to always guess correctly.

Instead of forcing all clients to configure guessing character encoding, it is much more efficient to require the sender to encode its messages correctly.

**Comment by progval** [comment]
> They send data that the client does not know how to display.

So does about every IRCd but Ergo. https://ircv3.net/specs/extensions/utf8-only aims to solve it, but it is not widely implemented yet.
